### PR TITLE
do not vote on own weights

### PR DESCRIPTION
**Pull request by n-hutton**
Looks like: 

image

### CODE REVIEW

File: colearn/ml_interface.py [hunk 1]
```diff
@@ -29,7 +29,7 @@ class ProposedWeights(BaseModel):
     weights: Weights
     vote_score: float
     test_score: float
-    vote: bool
+    vote: Optional[bool]
```
Comment (n-hutton) [inline colearn/ml_interface.py:32]: None indicates own vote/invalid

File: colearn/standalone_driver.py [hunk 2]
```diff
@@ -27,16 +27,23 @@ def run(n_rounds: int, learners: List[MachineLearningInterface]):
...
 
     vote = False
-    if approves >= len(learners) * vote_threshold:
+    if approves >= (len(prop_weights_list) - 1) * vote_threshold:
         vote = True
-        for j, learner in enumerate(learners):
-            learner.mli_accept_weights(prop_weights_list[j].weights)
+        # Set all learners to new weights
+        for _, learner in enumerate(learners):
+            learner.mli_accept_weights(prop_weights_list[0].weights)
```
Comment (evsmithx) [inline colearn/standalone_driver.py:47]: Minor, but this code would be equivalent and clearer if `new_weights` replaced `prop_weights_list[0].weights`. Also if you're not using the loop index then you can get rid of enumerate().

File: colearn/utils/plot.py [hunk 3]
```diff
@@ -110,12 +110,17 @@ def plot_votes(self, results: Results):
...
+
+        # Default fill: grey colour. Green and red will not match to 'None' in the votes_array
         coloured_votes_array[votes_array == 1] = green_colour
         coloured_votes_array[votes_array == 0] = red_colour
+        coloured_votes_array[(votes_array != 1) & (votes_array != 0)] = grey_colour
 
         # make extra legend entries
         red_patch_handle = mpatches.Patch(color=red_colour / 256, label='Negative vote')
         green_patch_handle = mpatches.Patch(color=green_colour / 256, label='Positive vote')
+        grey_patch_handle = mpatches.Patch(color=grey_colour / 256, label='N/A vote')
```
Comment (evsmithx) [inline colearn/utils/plot.py:123]: "Did not vote" would be a better label.

**Review by evsmithx — CHANGES_REQUESTED** [review]
I think the gRPC interface will be broken by this because it has vote: bool. Can you modify that as well?

File: colearn_grpc/grpc_learner_server.py [hunk 4]
```diff
@@ -181,7 +181,8 @@ def TestWeights(self, request_iterator, context):
             proposed_weights = self.learner.mli_test_weights(weights)
             pw.vote_score = proposed_weights.vote_score
             pw.test_score = proposed_weights.test_score
-            pw.vote = proposed_weights.vote
+            if proposed_weights.vote is not None:
```
Comment (n-hutton) [inline colearn_grpc/grpc_learner_server.py:184]: grpc allows you to not specify values

File: colearn/standalone_driver.py (hunk 2 above)
Comment (evsmithx) [reply colearn/standalone_driver.py:47]: You marked the conversation as resolved but you haven't fixed it!

**Review by evsmithx — APPROVED** [review]

**Review by evsmithx — APPROVED** [review]

**Pull request merged**